controllers/service.go

```diff
@@ -10,67 +10,104 @@ import (
-					Protocol:   corev1.ProtocolTCP,
-					Port:       portWS,
-					TargetPort: intstr.FromString("ws"),
-				},
-				{
-					Name:       "http",
-					Protocol:   corev1.ProtocolTCP,
-					Port:       portHTTP,
-					TargetPort: intstr.FromString("http"),
-				},
-				{
-					Name:       "api-http",
-					Protocol:   corev1.ProtocolTCP,
-					Port:       portAPIHTTP,
-					TargetPort: intstr.FromString("api-http"),
-				},
-				{
-					Name:       "proxy-http",
-					Protocol:   corev1.ProtocolTCP,
-					Port:       portProxyHTTP,
-					TargetPort: intstr.FromString("proxy-http"),
-				},
-				{
-					Name:       "cluster-swarm",
-					Protocol:   corev1.ProtocolTCP,
-					Port:       portClusterSwarm,
-					TargetPort: intstr.FromString("cluster-swarm"),
-				},
+	svcName := "ipfs-cluster-internal-" + m.Name
+	expected := expectedService(
+		svcName,
+		m.Name,
+		m.Namespace,
+		"ClusterIP",
+		m.Annotations,
+		[]corev1.ServicePort{
+			{
+				Name:       nameIpfsHTTP,
+				Protocol:   corev1.ProtocolTCP,
+				Port:       portIpfsHTTP,
+				TargetPort: intstr.FromString(nameIpfsHTTP),
 			},
-			Selector: map[string]string{
-				"app.kubernetes.io/name": "ipfs-cluster-" + m.Name,
+			{
+				Name:       nameClusterAPI,
+				Protocol:   corev1.ProtocolTCP,
+				Port:       portClusterAPI,
+				TargetPort: intstr.FromString(nameClusterAPI),
+			},
+			{
+				Name:       nameClusterProxy,
+				Protocol:   corev1.ProtocolTCP,
+				Port:       portClusterProxy,
+				TargetPort: intstr.FromString(nameClusterProxy),
+			},
+		},
+	)
+
+	expected.DeepCopyInto(svc)
+	// FIXME: catch this error before we run the function being returned
+	if err := ctrl.SetControllerReference(m, svc, r.Scheme); err != nil {
+		return func() error { return err }, ""
+	}
+	return func() error {
+		svc.Spec = expected.Spec
+		return nil
+	}, svcName
+}
+
+// If enabled, IPFS gateway serivce
+func (r *IpfsClusterReconciler) serviceGateway(
+	m *clusterv1alpha1.IpfsCluster,
+	svc *corev1.Service,
+) (controllerutil.MutateFn, string) {
+	svcName := "ipfs-cluster-gateway-" + m.Name
+	expected := expectedService(
+		svcName,
+		m.Name,
+		m.Namespace,
+		"LoadBalancer",
+		m.Annotations,
+		[]corev1.ServicePort{
+			{
+				Name:       nameIpfsHTTP,
+				Protocol:   corev1.ProtocolTCP,
+				Port:       portIpfsHTTP,
+				TargetPort: intstr.FromString(nameIpfsHTTP),
 			},
 		},
+	)
+	expected.DeepCopyInto(svc)
+	// FIXME: catch this error before we run the function being returned
+	if err := ctrl.SetControllerReference(m, svc, r.Scheme); err != nil {
+		return func() error { return err }, ""
 	}
+	return func() error {
+		svc.Spec = expected.Spec
+		return nil
+	}, svcName
+}
+
+// If enabled, the cluster API
+func (r *IpfsClusterReconciler) serviceAPI(
+	m *clusterv1alpha1.IpfsCluster,
+	svc *corev1.Service,
+) (controllerutil.MutateFn, string) {
+	svcName := "ipfs-cluster-api-" + m.Name
+	expected := expectedService(
+		svcName,
+		m.Name,
+		m.Namespace,
+		"LoadBalancer",
```

Comment:
Does this always need to be a loadbalancer? Could a ClusterIP not fulfill this role?